Adds new require-optimization rule.

EvNaverniouk opened the pull request:
Fixes #240. Adds a new rule called `require-optimization` which ensures all components either have a `shouldComponentUpdate` method or a `PureRenderMixin`.

Tested on several large React project repos without any issues using a beta release of this rule here (https://travis-ci.org/Eight-Solutions/eslint-plugin-cumul8). General feedback from users of this rule so far have been very positive.

ljharb commented:
Thanks! Can you add some tests that ensure that SFCs (both normal function and arrow function ones) do *not* warn with this rule?

EvNaverniouk commented:
@ljharb Ah, of course. Forgot about SFCs. Updated rule, added tests here: https://github.com/yannickcr/eslint-plugin-react/pull/601/commits/94fa138d138a528dc43d46b9a791a72ce0544db6

EvNaverniouk commented:
Actually, now that I think about it, maybe ignoring SFCs should be done behind an option flag. Since, technically, SFCs are **not** optimized components so they **should** throw an error. But at the moment, there is no way to optimize them. So I can imagine some authors might want to ban the user of SFCs for this reason.

What do you think?

ljharb commented:
I think throwing an error on all SFCs, especially when other rules exist to require SFCs whenever possible, is heavy handed enough that it would need to be defaulted off, or omitted entirely.

SFCs don't _need_ optimization, that's the point.

EvNaverniouk commented:
Alright, I'll leave the rule as is then. Ready to merge?

ljharb commented:
LGTM, but @yannickcr should take a look just to be sure.

What about a shouldComponentUpdate that's the default, i.e., `return true`?

yannickcr commented:
LGTM too, great work!
